openhands/storage/batched_web_hook.py

```diff
@@ -169,38 +168,42 @@ def _queue_update(
                 self._batch_timer = None
 
             timer = threading.Timer(
-                self.batch_timeout_seconds, self._send_batch_from_timer
+                self.batch_timeout_seconds, self._enqueue_batch_from_timer
             )
             timer.daemon = True
             timer.start()
             self._batch_timer = timer
 
-    def _send_batch_from_timer(self) -> None:
+    def _enqueue_batch_from_timer(self) -> None:
         """Send the batch from the timer thread.
         This method is called by the timer and submits the actual sending to the executor.
         """
-        EXECUTOR.submit(self._send_batch)
+        with self._batch_lock:
+            self._enqueue_batch_from_lock()
 
-    def _send_batch(self) -> None:
+    def _enqueue_batch_from_lock(self, background=True) -> None:
+        """
+        Must have lock before calling. Will reset the batch state and send the current one.
+        Uses executor by default, but can perform synchronously by setting background=False
+        """
+        batch_to_send = self._batch
+        self._batch = {}
+        self._batch_size = 0
+        if background:
+            EXECUTOR.submit(self._send_batch, batch_to_send)
+        else:
+            self._send_batch(batch_to_send)
+        # Cancel any pending timer
+        if self._batch_timer is not None:
+            self._batch_timer.cancel()
+            self._batch_timer = None
+
+    def _send_batch(
+        self, batch_to_send: dict[str, tuple[str, Optional[Union[str, bytes]]]]
```

Comment:
```suggestion
        self, batch_to_send: dict[str, tuple[str, str | bytes | None]]
```

NIT: use modern typing syntax